### ISSUE TITLE
Refactor Bar and Box Plots

### ISSUE THREAD

**Issue by swharden** [swharden/ScottPlot#315]
The current bar plots work somewhat, but it's not easy to add new features to them. The bar plot module should be refactored to accomodate several new desired features. This issue unifies #24, #260, and #277. After this module is complete, it can probably be modified slightly to support box plots too (#244) to make it easier to display grouped box plots. Perhaps a single class can be used for bar and box plots.

### Example (from matplotlib)

![image](https://user-images.githubusercontent.com/4165489/77127431-cda81b00-6a22-11ea-8a14-c7c6c676f83d.png)

* This graph has 5 ***groups*** (G1, G2, G3, G4, and G5)
* This graph has 2 ***bar series*** (Men and Women)

One can say that if several ***bar series*** objects have the same number of elements in them, it is possible to display them organized by ***group***. Each bar series and each group has its own label.

### Features:
- [x] can display a single bar series
- [x] can display several bar series arranged by groups
- [x] each bar series can be individually customized
- [x] each bar series should appear separately in the legend
- [x] it should be easy to display as vertical or horizontal bars
- [x] it should be possible to display as stacked bars
- [x] optional error-bars should be supported
- [x] add option to customize lines
- [x] add option to display values over bars
- [x] add support for box plots
- [x] display scatter data next to plots
- [x] display data distribution curve

**Comment by swharden** [comment]
Experimental code is in:
https://github.com/swharden/ScottPlot/blob/master/tests/Experimental/BarGraphTests.cs

### Single Bar Series

```cs
double[,] values = new double[,] { { 7, 12, 40, 40, 100, 125, 172, 550, 560, 600, 2496, 2789 } };
string[] rowLabels = new string[] { null };
string[] colLabels = new string[] { "ant", "bird", "mouse", "human", "cat", "dog", "frog", "lion", "elephant", "horse", "shark", "hippo" };

var plt = new ScottPlot.Plot(400, 300);

// create and add an experimental plottable (not in the ScottPlot namespace)
var pltBar = new PlottableBar2(rowLabels, colLabels, values);
plt.Add(pltBar);

plt.Title("Body-to-Brain Mass Ratio");
plt.XTicks(colLabels);
plt.Axis(-1, colLabels.Length, 0, null);
plt.Grid(enableVertical: false);

// customize ticks to display better at high density
plt.Ticks(useMultiplierNotation: false);
plt.Ticks(xTickRotation: 45);
plt.Layout(xLabelHeight: 40);
```

![image](https://user-images.githubusercontent.com/4165489/77128533-cedb4700-6a26-11ea-9b41-36eaa250c047.png)

### Multiple Bar Series (grouped)

```cs
double[,] values = new double[,] { { 15, 22, 45, 17 }, { 37, 21, 29, 13 } };
string[] rowLabels = new string[] { "men", "women" };
string[] colLabels = new string[] { "always", "regularly", "sometimes", "never" };

var plt = new ScottPlot.Plot(400, 300);

var pltBar = new PlottableBar2(rowLabels, colLabels, values);
plt.Add(pltBar);

plt.Title("How often do you read reviews?");
plt.XTicks(colLabels);
plt.Ticks(useMultiplierNotation: false);
plt.Axis(-1, colLabels.Length, 0, null);
plt.Grid(enableVertical: false);
```

![image](https://user-images.githubusercontent.com/4165489/77128495-ae12f180-6a26-11ea-9775-ce358cf5d4e5.png)

### Remaining Problem / Refactor Required

I like the direction this is going, but there may be utility in making each bar series its own plottable rather than grouping them into a 2d data array. The major advantage is that each bar series could easily appear as its own item in the legend. This gets complicated if the group of multiple bar series (the current implementation) is a single plottable - it wants to render in the legend as a single thing.

Breaking it up into individual `PlottableBarSeries` objects would also make styling of individual bars easier.

**Comment by swharden** [comment]
Here's what I came up with. The legend module doesn't support it yet, but I think the way to solve that issue is to have something like an `IMultiLabel` that some plottables have, then the legend knows to display multiple labels from a single plottable

I like this because it doesn't require the user to make 2D arrays. It also clearly separates construction of bar series from bar groups.

```cs
// define values, labels, and style of each bar series
var barSetMen = new BarSet(new double[] { 15, 22, 45, 17 }, "Men");
var barSetWomen = new BarSet(new double[] { 37, 21, 29, 13 }, "Women");

// collect BarSets into groups
var barSets = new BarSet[] { barSetMen, barSetWomen };
var groupLabels = new string[] { "always", "regularly", "sometimes", "never" };

// create the experimental plottable
var plottableThing = new PlottableBar3(barSets, groupLabels);
Console.WriteLine(plottableThing);

// plot the experimental plottable
var plt = new ScottPlot.Plot(400, 300);
plt.Add(plottableThing);
plt.Title("How often do you read reviews?");
plt.XTicks(groupLabels);
plt.Ticks(useMultiplierNotation: false);
plt.Grid(enableVertical: false);
TestTools.SaveFig(plt);
```

![image](https://user-images.githubusercontent.com/4165489/77227127-4a86e200-6b54-11ea-8dfe-0304c373c651.png)

**Comment by swharden** [comment]
I think this new box plot is similar/better to the current one, and it should probably replace it in the next NuGet release. I will leave this issue open and keep working on adding more features (like horizontal mode)...

### Single Series
```cs
var plt = new ScottPlot.Plot(600, 400);

// define values, labels, and style of each bar series
var barSetBBR = new DataSet("brainBodyRatio", new double[] { 7, 12, 40, 40, 100, 125, 172, 550, 560, 600, 2496, 2789 });

// collect BarSets into groups
var datasets = new DataSet[] { barSetBBR };
var groupLabels = new string[] { "ant", "bird", "mouse", "human", "cat", "dog", "frog", "lion", "elephant", "horse", "shark", "hippo" };

// create the experimental plottable
var plottableThing = new PlottableBarExperimental(datasets, groupLabels);

// plot the experimental plottable
plt.Add(plottableThing);
plt.Title("Body-to-Brain Mass Ratio");
plt.XTicks(groupLabels);
plt.Axis(null, null, 0, null);
plt.Ticks(useMultiplierNotation: false);
plt.Grid(enableVertical: false);
plt.SaveFig("Experimental_BarExperimental_Bar3Single.png");
```
![image](https://user-images.githubusercontent.com/4165489/77499793-742c5b80-6e29-11ea-8648-2e69b1455693.png)

### Multi-Series
```cs
var plt = new ScottPlot.Plot(600, 400);

// define values, labels, and style of each bar series
var barSetMen = new DataSet("Men", new double[] { 15, 22, 45, 17 });
var barSetWomen = new DataSet("Women", new double[] { 37, 21, 29, 13 });

// collect BarSets into groups
var datasets = new DataSet[] { barSetMen, barSetWomen };
var groupLabels = new string[] { "always", "regularly", "sometimes", "never" };

// create the experimental plottable
var plottableThing = new PlottableBarExperimental(datasets, groupLabels, stacked: true);

// plot the experimental plottable
plt.Add(plottableThing);
plt.Legend(location: legendLocation.upperRight);
plt.Title("How often do you read reviews?");
plt.YLabel("Respondents (%)");
plt.XTicks(groupLabels);
plt.Axis(null, null, 0, null);
plt.Grid(enableVertical: false);
plt.SaveFig("Experimental_BarExperimental_Bar3.png");
```

![image](https://user-images.githubusercontent.com/4165489/77499865-adfd6200-6e29-11ea-96df-5ab10b8a408c.png)

### Multi-Series (Stacked)

```cs
var plt = new ScottPlot.Plot(600, 400);
// define values, labels, and style of each bar series
var barSetMen = new DataSet("Men", new double[] { 15, 22, 45, 17 });
var barSetWomen = new DataSet("Women", new double[] { 37, 21, 29, 13 });

// collect BarSets into groups
var datasets = new DataSet[] { barSetMen, barSetWomen };
var groupLabels = new string[] { "always", "regularly", "sometimes", "never" };

// create the experimental plottable
var plottableThing = new PlottableBarExperimental(datasets, groupLabels, stacked: true);

// plot the experimental plottable
plt.Add(plottableThing);
plt.Legend(location: legendLocation.upperRight);
plt.Title("How often do you read reviews?");
plt.YLabel("Respondents (%)");
plt.XTicks(groupLabels);
plt.Axis(null, null, 0, null);
plt.Grid(enableVertical: false);
plt.SaveFig("Experimental_BarExperimental_Bar3Stacked.png");
```

![image](https://user-images.githubusercontent.com/4165489/77499879-b786ca00-6e29-11ea-8cee-40b64e5b6eee.png)

**Comment by swharden** [comment]
Wow! Rewriting the vertical rendering functions to render horizontal bar graphs was so much harder than I thought it would be... I think my brain rotated 90º inside my skull 😵 but the output is great now!

### Defining Data Sets & Groups
```cs
// define series-level data
var barSetMen = new DataSet(label: "Men",
	values: new double[] { 15, 22, 45, 17 },
	errors: new double[] { 6, 3, 8, 4 });

var barSetWomen = new DataSet(label: "Women",
	values: new double[] { 37, 21, 29, 13 },
	errors: new double[] { 7, 5, 6, 3 });

// define group-level data
var dataSets = new DataSet[] { barSetMen, barSetWomen };
var groupLabels = new string[] {
	"Always", "Regularly", "Sometimes", "Never" };
```

### Vertical Grouped Bar Graph
```cs
plt.PlotBar(dataSets, groupLabels);
```
![image](https://user-images.githubusercontent.com/4165489/77603572-d008ea00-6ee6-11ea-88b9-c1a5c79505af.png)

### Horizontal Grouped Bar Graph
```cs
plt.PlotBar(dataSets, groupLabels, horizontal: true);
```
![image](https://user-images.githubusercontent.com/4165489/77603587-d6976180-6ee6-11ea-8762-f3d109fe38c5.png)

### Vertical Stacked Bar Graph
```cs
plt.PlotBar(dataSets, groupLabels, stacked: true);
```
![image](https://user-images.githubusercontent.com/4165489/77603667-07779680-6ee7-11ea-82ba-fd19b8e0f390.png)

### Horizontal Stacked Bar Graph
```cs
plt.PlotBar(dataSets, groupLabels, stacked: true, horizontal: true);
```
![image](https://user-images.githubusercontent.com/4165489/77603709-2b3adc80-6ee7-11ea-95d0-dd9ecabdfeab.png)

**Comment by swharden** [comment]
I keep getting stuck not liking the architecture here. Experimenting with a refactor https://github.com/swharden/ScottPlot/tree/bar-refactor I realized that this task can be simplified if each bar is its own object and has the ability to render itself (horizontally or vertically). Organization (grouping bars into bar series, and grouping those into a multi-series object) is a lot easier too.

There are strong implications for whisker plots #246 #247 and if this "multi-series" organization is handled right, one object can be used to display data as box plots and bar graphs. 

### Example of desired features (from ggplot)
* there are 4 groups: a, b, c, d
* there are 2 series: one, two
* data points can be displayed over (or beside) box/bar

![image](https://user-images.githubusercontent.com/4165489/77719538-e0d75f80-6fbb-11ea-80bc-a3518898a776.png)

### Language from GraphPad
Maybe I can mimic established software for the the user-facing components. GraphPad has a [how to make a grouped box and whiskers plot](https://www.graphpad.com/support/faq/graph-tip-how-to-make-a-grouped-box-and-whiskers-plot/) tutorial which can be an inspiration.

![image](https://user-images.githubusercontent.com/4165489/77719679-41ff3300-6fbc-11ea-9c79-0684c509046c.png)

---

still working on this

**Comment by swharden** [comment]
I found an architecture I think I like now.

(this is fake data)

![image](https://user-images.githubusercontent.com/4165489/77836081-5c0a5400-7129-11ea-9904-b6ce25748445.png)

The plottable for this is actually pretty simple. It holds `Population` (previously `PopulationStats`) objects and keeps track of their grouping, but it doesn't render them.
https://github.com/swharden/ScottPlot/blob/0482a68455511d1ba9cb56a28267b39f32d8616a/src/ScottPlot/plottables/PlottableGroupedSeries.cs#L1-L89

A dedicated static class handles rendering of populations. It can render as scatter, box, bar, etc.

https://github.com/swharden/ScottPlot/blob/0482a68455511d1ba9cb56a28267b39f32d8616a/src/ScottPlot/plottables/RenderPopulation.cs#L1-L174

**Comment by swharden** [comment]
I think all the parts are here now. It just needs an intelligent function in the Plot module. This is surprisingly complicated because there are so many options 🤔 

### Single Series: `PopulationSeries` = `Population[]`
![image](https://user-images.githubusercontent.com/4165489/77854203-d0d79f80-71b6-11ea-9692-f3f340b4055b.png)

### Multi-Series: `PopulationMultiSeries` = `PopulationSeries[]`
![image](https://user-images.githubusercontent.com/4165489/77854212-d9c87100-71b6-11ea-8cd7-d632a5a94648.png)

**Comment by swharden** [comment]
I think this is in a good place now. This got released in ScottPlot 4.0.25. Check out the cookbooks to see what's possible now:

### Bar Plot
This is a relatively simple plot type. It's just for bar plots, and it does not display population details (like scatter plots of individual data points).
http://swharden.com/scottplot/cookbooks/4.0.25/#PlotTypes_Bar_Quickstart

### Box and Whisker Plot
This has extensive options, including displaying as bar plots, and rendering population data points next to the box/bar plot.
http://swharden.com/scottplot/cookbooks/4.0.25/#PlotTypes_Populations_PlotPopulationSeriesUnique
